augur/augur_client.py

```diff
@@ -150,58 +158,40 @@ def event_signature_to_name_map(self):
-      IOError: If there is an issue connecting to the Augur node.
-    '''
-    self._require_is_open()
-    tasks = [self._start_new_block_listener(),
-             self._fetch_historical_events(event_name_list, current_block,
-                                           on_event, increment_block)]
-    # Terminate tasks once the historical listener has completed.
-    _, pending = await asyncio.wait(tasks, return_when=asyncio.FIRST_COMPLETED)
-    for task in pending:
-      task.cancel()
-
-  async def start_event_listener(self, event_name_list, on_event):
-    '''Polls for new Augur events.
+  def filter_blocks(self, start_block, end_block, filter, handler,
+      increment=1000):
+    '''Retrieve all event logs from the start_block to the end_block. Once the
+    end block has been reached, filter_blocks will start a realtime listener to
+    run the handler everytime a new event occurs.
 
     Args:
-      event_name_list (list[string]): List containing names of Augur events.
-      on_event (callback function): Function to run when an Augur event occurs.
+      start_block (int): Starting block number.
+      end_block (int): Ending block number. Can also be the string 'latest'
+                       which represents the latest Ethereum block.
+      filter (dict): Dictionary containing filter parameters.
+      handler (callback): Callback function for when an event log is found in
+                          the current block.
+      increment (int): Specifies the range of blocks from the current_block to
+                       look for Augur logs.
 
-    Raise:
-      IOError: If there is an issue connecting to the Augur node.
+    Raises:
+      IOError: If there is an issue with communicating with the node.
     '''
     self._require_is_open()
-    ethereum_uri = self._ethereum_client.providers[0].endpoint_uri
-    websocket = await websockets.connect(ethereum_uri)
-    topic_list = [self._event_name_to_signature_map[event_name]
-                  for event_name in event_name_list]
-    address = self._ethereum_client.toChecksumAddress(self._addresses['Augur'])
-    await self._send_request('eth_subscribe', websocket, [
-      'logs', {
-        'address': address,
-        'topics': [topic_list],
-        'fromBlock': 'latest'
-      }])
-    async for event in websocket:
-      event_json = json.loads(event)['params']['result']
-      on_event(event_json)
+    filter['address'] = self._ethereum_client.toChecksumAddress(
+      self._addresses['Augur'])
+    for current_block in range(start_block, end_block, increment):
+      filter['fromBlock'] = current_block
+      filter['toBlock'] = current_block + increment
+      event_filter = self._ethereum_client.eth.filter(filter)
+      for event in event_filter.get_all_entries():
+        handler(event)
+    loop = asyncio.get_event_loop()
+    filter['fromBlock'] = hex(end_block)
+    filter.pop('toBlock', None)
+    ttt = threading.Thread(target=self._start_realtime_listener,
```

Comment:
Try to do this without using another thread. Instead get the current thread's event loop and schedule to run ```_realtime_listener``` at some future point in time using ```asyncio.create_task```.

Reply:
It looks like uvicorn destroys the asyncio event loop and starts a new one. We can tell uvicorn to use the same event loop that has the `filter_blocks` task on it by passing it to the `run` method:
```python
uvicorn.run(app, .., loop=asyncio.get_event_loop())
```

Reply:
Well man that would be awesome. See if it works and thanks for looking into that.